sysml.library/Kernel Library/Occurrences.kerml

```diff
@@ -1,297 +1,561 @@
+	 * assert complete inclusion and exclusion in time or space, or both, which includes
+	 * portions of an occurrence (having the same identity).  Portions include slices and shots
+	 * over time and space.
+	 */
+	abstract class Occurrence specializes Anything {
+		private import SequenceFunctions::*;
+		
+		feature portionOfLife: Life[1] subsets portionOf;
+		
+		feature self: Occurrence[1] redefines Anything::self subsets timeSlices, spaceSlices, sameLifeOccurrences;
+		feature sameLifeOccurrences: Occurrence[1..*] subsets things;
+		
+		/**
+		 * Occurrences that end before this occurrence starts.
+		 */
+		feature predecessors: Occurrence[0..*] subsets occurrences;
+		
+		/**
+		 * Occurrences that start after this occcurrence ends.
+		 */
+		feature successors: Occurrence[0..*] subsets occurrences;
+		
+		/**
+		 * Occurrences that end just before this occurrence starts, with no
+		 * possibility of other occurrences happening in the time between them.
+		 */
+		feature immediatePredecessors: Occurrence[0..*] subsets predecessors;
+
+		/**
+		 * Occurrences that start just after this occurrence ends, with no
+		 * possibility of other occurrences happening in the time between them.
+		 */
+		feature immediateSuccessors: Occurrence[0..*] subsets successors;
+
+		/**
+		 * Occurrences that start no earlier than and end no later than
+		 * this occurrence, including at least this occurrence.
+		 */
+		feature timeEnclosedOccurrences: Occurrence[1..*] subsets occurrences;
+
+		/**
+		 * Occurrences that this one completely includes in space (not necessarily in time),
+		 * including this one.
+		 */
+		feature spaceEnclosedOccurrences: Occurrence[1..*] subsets occurrences;
+
+		/**
+		 * Occurrences that this one completely includes in both space and time,
+		 * including this one.
+		 */
+		feature spaceTimeEnclosedOccurrences: Occurrence[1..*] subsets timeEnclosedOccurrences, spaceEnclosedOccurrences;
+
+		/**
+		 * All space time enclosed occurrences that take up zero time and space.
+		 */
+		feature all spaceTimeEnclosedPoints : Occurrence[1..*] subsets spaceTimeEnclosedOccurrences {
+			redefines innerSpaceDimension = 0;
+			binding startShot[1] = endShot[1];
+		}
+
+		/**
+		 * The number of variables need to identify space points in this occurrence, from 0
+		 * to 3, without regard to higher dimensional spaces it might be emedded in.
+		 */
+		feature innerSpaceDimension : Integer [1];
+		inv { innerSpaceDimension >= 0 & innerSpaceDimension <= 3 }
+
+		/**
+		 * For occurrences of innerSpaceDimension 1 or 2, the number of variables need to
+		 * identify their space points in higher dimensions they might be embedded in, from
+		 * the innerSpaceDimension to 3. An outerSpaceDimension equal to innerSpaceDimension
+		 * indicates the occurrence is spatially straight (innerSpaceDimension 1 embedded in
+		 * 2 or 3 dimensions) or flat (innerSpaceDimension 2 embedded in 3 dimensions).
+		 */
+		feature outerSpaceDimension : Integer [0..1];

```

Comment:
Can be typed `Natural[1]` as dimensions should never be negative.

Reply:
Done.